### PR TITLE
feat: support static filter for target table in merge into statement

### PR DESCRIPTION
**Pull request by SkyFan2002**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

- Closes #issue

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13426)

### CODE REVIEW

**Comment by dantengsky** [comment]
askbend:summary

**Comment by JackTan25** [comment]
This is a good pr, but we need to do test,  please test merge into,thanks!

**Comment by dantengsky** [comment]
> This is a good pr, but we need to do test, please test merge into,thanks!

seems there is a test case for merge into

https://github.com/datafuselabs/databend/pull/13426/files#diff-c4ebd79f73c589be0b82530a758d6ffbcabda5fd530df8e0a1877d0685c318e9R45

~~~
statement ok
merge into t1 using (select * from t2 ) as t2 on t1.a = t2.a  when matched then update set t1.c = t2.c;
~~~

**Comment by dantengsky** [comment]
@SkyFan2002  test asserts `system.metrics` seems flaky, consider removing it?

https://github.com/datafuselabs/databend/actions/runs/6650877640/job/18072111288?pr=13426#step:4:296

~~~
Error: SelfError("sqllogictest failed")
0: query result mismatch:
[SQL] select value from system.metrics where metric = 'fuse_segments_range_pruning_before_total';
[Diff] (-expected|+actual)
-   4.0
+   8.0
at tests/sqllogictests/suites/mode/standalone/metrics.test:47
~~~

File: src/query/service/src/interpreters/interpreter_merge_into.rs [hunk 1]
```diff
@@ -343,6 +367,223 @@ impl MergeIntoInterpreter {
...
+        let source_plan = join.child(0)?;
+        let join_op = match join.plan() {
+            RelOperator::Join(j) => j,
+            _ => unreachable!(),
+        };
+        if join_op.left_conditions.len() != 1 || join_op.right_conditions.len() != 1 {
+            return Ok(Box::new(join.clone()));
+        }
+        let source_side_expr = &join_op.left_conditions[0];
+        let target_side_expr = &join_op.right_conditions[0];
```
Comment (JackTan25) [inline src/query/service/src/interpreters/interpreter_merge_into.rs:395]: why just use only one condition?

**Review by dantengsky — APPROVED** [review]

**Review by JackTan25 — APPROVED** [review]
LGTM, this pr is great, let improve more step by step, thanks. cc @SkyFan2002 @dantengsky

File: src/query/service/src/interpreters/interpreter_merge_into.rs (hunk 1 above)
Comment (SkyFan2002) [reply src/query/service/src/interpreters/interpreter_merge_into.rs:395]: Yes, let's continue improving this in later PR

**Comment by SkyFan2002** [comment]
Tested 3000 times in standalone mode: cc @JackTan25 
```
[2023-10-26T11:56:22Z INFO  test_replace_recluster] ==========================
[2023-10-26T11:56:22Z INFO  test_replace_recluster] ====verify table state====
[2023-10-26T11:56:22Z INFO  test_replace_recluster] ==========================
[2023-10-26T11:56:22Z INFO  test_replace_recluster]                            
[2023-10-26T11:56:22Z INFO  test_replace_recluster]                            
[2023-10-26T11:56:22Z INFO  test_replace_recluster] number of successfully executed merge-into statements : 3000
[2023-10-26T11:56:22Z INFO  test_replace_recluster]                            
[2023-10-26T11:56:22Z INFO  test_replace_recluster]                            
[2023-10-26T11:56:22Z INFO  test_replace_recluster] CHECK: value of successfully executed merge-into statements
[2023-10-26T11:56:22Z INFO  test_replace_recluster] CHECK: value of successfully executed merge-into statements: client 3000000, server 3000000
[2023-10-26T11:56:22Z INFO  test_replace_recluster] CHECK: distinct ids: client 3000, server 3000
[2023-10-26T11:56:22Z INFO  test_replace_recluster] CHECK: value of correlated column
[2023-10-26T11:56:22Z INFO  test_replace_recluster] CHECK: full table scanning
[2023-10-26T11:56:22Z INFO  test_replace_recluster] ===========================
[2023-10-26T11:56:22Z INFO  test_replace_recluster] ======     PASSED      ====
[2023-10-26T11:56:22Z INFO  test_replace_recluster] ===========================
[2023-10-26T11:56:22Z INFO  test_replace_recluster]                            
[2023-10-26T11:56:22Z INFO  test_replace_recluster]                            
[2023-10-26T11:56:22Z INFO  test_replace_recluster] ========METRICS============
[2023-10-26T11:56:22Z INFO  test_replace_recluster] fuse_commit_mutation_unresolvable_conflict_total : 2042.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] fuse_remote_io_read_bytes_after_merged_total : 49877745900.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] fuse_remote_io_seeks_after_merged_total : 54324.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_accumulate_milliseconds : [{"less_than":10.0,"count":1946.0},{"less_than":50.0,"count":1946.0},{"less_than":100.0,"count":1946.0},{"less_than":250.0,"count":1946.0},{"less_than":500.0,"count":1946.0},{"less_than":1000.0,"count":1946.0},{"less_than":2500.0,"count":1946.0},{"less_than":5000.0,"count":1946.0},{"less_than":10000.0,"count":1946.0},{"less_than":20000.0,"count":1946.0},{"less_than":30000.0,"count":1946.0},{"less_than":60000.0,"count":1946.0},{"less_than":300000.0,"count":1946.0},{"less_than":600000.0,"count":1946.0},{"less_than":1800000.0,"count":1946.0},{"less_than":null,"count":1946.0}]
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_accumulate_milliseconds_count : 1946.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_accumulate_milliseconds_sum : 1.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_append_blocks_counter_total : 4946.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_append_blocks_rows_counter_total : 4284000.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_apply_milliseconds : [{"less_than":10.0,"count":3000.0},{"less_than":50.0,"count":3002.0},{"less_than":100.0,"count":3005.0},{"less_than":250.0,"count":3013.0},{"less_than":500.0,"count":3030.0},{"less_than":1000.0,"count":3071.0},{"less_than":2500.0,"count":3424.0},{"less_than":5000.0,"count":3428.0},{"less_than":10000.0,"count":3428.0},{"less_than":20000.0,"count":3428.0},{"less_than":30000.0,"count":3428.0},{"less_than":60000.0,"count":3428.0},{"less_than":300000.0,"count":3428.0},{"less_than":600000.0,"count":3428.0},{"less_than":1800000.0,"count":3428.0},{"less_than":null,"count":3428.0}]
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_apply_milliseconds_count : 3428.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_apply_milliseconds_sum : 577127.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_deleted_blocks_counter_total : 81.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_deleted_blocks_rows_counter_total : 81000.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_matched_operation_milliseconds : [{"less_than":10.0,"count":1946.0},{"less_than":50.0,"count":1946.0},{"less_than":100.0,"count":1946.0},{"less_than":250.0,"count":1946.0},{"less_than":500.0,"count":1946.0},{"less_than":1000.0,"count":1946.0},{"less_than":2500.0,"count":1946.0},{"less_than":5000.0,"count":1946.0},{"less_than":10000.0,"count":1946.0},{"less_than":20000.0,"count":1946.0},{"less_than":30000.0,"count":1946.0},{"less_than":60000.0,"count":1946.0},{"less_than":300000.0,"count":1946.0},{"less_than":600000.0,"count":1946.0},{"less_than":1800000.0,"count":1946.0},{"less_than":null,"count":1946.0}]
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_matched_operation_milliseconds_count : 1946.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_matched_operation_milliseconds_sum : 2427.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_matched_rows_total : 1284000.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_not_matched_operation_milliseconds : [{"less_than":10.0,"count":3000.0},{"less_than":50.0,"count":3000.0},{"less_than":100.0,"count":3000.0},{"less_than":250.0,"count":3000.0},{"less_than":500.0,"count":3000.0},{"less_than":1000.0,"count":3000.0},{"less_than":2500.0,"count":3000.0},{"less_than":5000.0,"count":3000.0},{"less_than":10000.0,"count":3000.0},{"less_than":20000.0,"count":3000.0},{"less_than":30000.0,"count":3000.0},{"less_than":60000.0,"count":3000.0},{"less_than":300000.0,"count":3000.0},{"less_than":600000.0,"count":3000.0},{"less_than":1800000.0,"count":3000.0},{"less_than":null,"count":3000.0}]
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_not_matched_operation_milliseconds_count : 3000.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_not_matched_operation_milliseconds_sum : 0.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_replace_blocks_counter_total : 1865.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_replace_blocks_rows_counter_total : 477821193.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_split_milliseconds : [{"less_than":10.0,"count":4946.0},{"less_than":50.0,"count":4946.0},{"less_than":100.0,"count":4946.0},{"less_than":250.0,"count":4946.0},{"less_than":500.0,"count":4946.0},{"less_than":1000.0,"count":4946.0},{"less_than":2500.0,"count":4946.0},{"less_than":5000.0,"count":4946.0},{"less_than":10000.0,"count":4946.0},{"less_than":20000.0,"count":4946.0},{"less_than":30000.0,"count":4946.0},{"less_than":60000.0,"count":4946.0},{"less_than":300000.0,"count":4946.0},{"less_than":600000.0,"count":4946.0},{"less_than":1800000.0,"count":4946.0},{"less_than":null,"count":4946.0}]
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_split_milliseconds_count : 4946.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_split_milliseconds_sum : 2.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] query_merge_into_unmatched_rows_total : 3000000.0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] ===========================
[2023-10-26T11:56:22Z INFO  test_replace_recluster]                            
[2023-10-26T11:56:22Z INFO  test_replace_recluster]                            
[2023-10-26T11:56:22Z INFO  test_replace_recluster] ======CLUSTERING INFO======
[2023-10-26T11:56:22Z INFO  test_replace_recluster] cluster_key : (to_yyyymmdd(insert_time), id)
[2023-10-26T11:56:22Z INFO  test_replace_recluster] block_count: 12
[2023-10-26T11:56:22Z INFO  test_replace_recluster] constant_block_count: 0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] unclustered_block_count: 0
[2023-10-26T11:56:22Z INFO  test_replace_recluster] average_overlaps: 7.5
[2023-10-26T11:56:22Z INFO  test_replace_recluster] average_depth: 6
[2023-10-26T11:56:22Z INFO  test_replace_recluster] block_depth_histogram: {"00006":12}
[2023-10-26T11:56:22Z INFO  test_replace_recluster] ===========================
```

**Pull request merged**